crates/sparrow-runtime/src/prepare/execute_iter.rs

```diff
@@ -0,0 +1,434 @@
+use std::pin::Pin;
+use std::sync::{Arc, PoisonError};
+use std::task::Poll;
+
+use arrow::array::{PrimitiveArray, TimestampNanosecondArray};
+use arrow::compute::SortColumn;
+use arrow::datatypes::{ArrowPrimitiveType, SchemaRef, TimestampNanosecondType, UInt64Type};
+use arrow::error::ArrowError;
+use arrow::record_batch::RecordBatch;
+use error_stack::{IntoReport, IntoReportCompat, Report, ResultExt};
+use futures::stream::BoxStream;
+use futures::{Stream, StreamExt};
+use itertools::Itertools;
+use sparrow_api::kaskada::v1alpha::{slice_plan, TableConfig};
+use sparrow_core::{downcast_primitive_array, TableSchema};
+
+use crate::execute::key_hash_inverse::ThreadSafeKeyHashInverse;
+use crate::prepare::slice_preparer::SlicePreparer;
+use crate::prepare::Error;
+use crate::RawMetadata;
+
+use super::column_behavior::ColumnBehavior;
+
+/// An iterator over prepare batches. TODO
+///
+/// In addition to iterating, this is responsible for the following:
+///
+/// 1. Inserting the time column, subsort column, and key hash from source to
+///    the batch
+/// 2. Casts required columns
+/// 3. Sorts the record batches by the time column, subsort column, and key hash
+pub struct ExecuteIter<'a> {
```

Comment:
Similar to the `PrepareIter`, but includes the input buffer / late data / watermark logic. 

I'll create a `StreamReader`, similar to the `TableReader`, that the `Scan` operation uses depending on what input type it's reading from. The `StreamReader` will initialize a `pulsar_stream` as the `reader`, then create an `ExecuteIter` to read batches from.

Reply:
Haven't looked at this (which seems to be the main thing in this PR). Happy to approve the other pieces in a separate PR, or look at this further in my next pass.

Reply:
Prefer looking at this in this PR, we can go over together if that's faster to follow up too